README.markdown
# Tempr
## a Ruby temporal expressions library


This is a work in progress. See date_time_range.rb for usage examples.


### License:

Copyright (c) 2012 Eric Gjertsen

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.markdown
# Changelog

## 0.1.4 / 2012-02-20

- add range adjacency methods (`precedes?`, `succeeds?`, `adjacent_to?`)

## 0.1.3 / 2012-02-19

- convert date to time ranges for date <-> time comparisons in `within?/subsume?/intersection_with`
- fix behavior with exclusive ranges
- add test cases, DRY up tests
- add Rakefile, add rake to gemspec

## 0.1.2 / 2012-02-19

- add `within?`, `subsume?`, `intersection_with`, `intersects?`

## 0.1.1 / 2012-02-18

- add `SubRangeIterator#cover?` shortcut
- add `between_times`
- `on_day` renamed `each_day_of_month`; deprecation warning
- `at_time` renamed `each_time_of_day`; deprecation warning
- add utc_offset param to `each_time_of_day`, `time_range`
- fix error where time subranges forced to local time

## 0.1.0 / 2012-02-16

- Initial gem release


test/suite.rb
%w[ time_subrange 
    each_time_of_day 
    range_within_other
    range_intersection
  ].each do |test|
      require File.expand_path(test,File.dirname(__FILE__))
    end

test/at_time.rb
require File.expand_path('test_helper', File.dirname(__FILE__))

module AtTimeTests
  module Fixtures

    BaseRanges = {:local => Time.parse('2012-01-01 00:00:00 -0500')...
                            Time.parse('2013-01-01 00:00:00 -0500'),
                  :utc   => Time.parse('2012-01-01 00:00:00 UTC')...
                            Time.parse('2013-01-01 00:00:00 UTC')
                 }
                
  end

  describe 'DateTimeRange#at_time' do

    [:local, :utc].each do |time_type|
    
      describe "across daylight savings time boundaries for #{time_type} times" do
      
        let(:subject) { Fixtures::BaseRanges[time_type].extend(Tempr::DateTimeRange) }
        
        it 'should be at the same time of day regardless of time zone' do
          subject.each_day.at_time("2:00pm",60*60).each do |range|
            offset = range.begin.utc_offset
            actual = range.begin.getlocal(offset).hour
            #puts "#{range}"
            assert_equal 14, actual, "for range: #{range}"
          end
        end
        
      end
      
    end
  end

end  # namespace

test/test_helper.rb
require File.expand_path('../lib/tempr', File.dirname(__FILE__))

gem "minitest" 

require 'minitest/spec'
MiniTest::Unit.autorun

scratch/chainable.rb
# first a simpler case: 
# subranges of integers

module Subrange

  def each_nth(n,len=1,offset=0)
    SubRangeIterator.new( self, n, len,
                          lambda {|dt| dt + offset},
                          lambda {|dt,i| dt + i}
                        )
  end  
  
  class SubRangeIterator
    include Enumerable
    include Subrange
    
    attr_accessor :range, :step, :length, :initial_proc, :step_proc

    def begin; self.range.begin; end
    def end; self.range.end; end
    
    def initialize(range, step, length, initial_proc, step_proc)
       self.range, self.step, self.length, self.initial_proc, self.step_proc = \
            range,      step,      length,      initial_proc,      step_proc
    end
  
    def each(&b)
      if SubRangeIterator === self.range
        self.range.each do |sub|
#          puts "sub.range = #{sub}"
          each_by_step(sub, &b)
        end
      else
        each_by_step do |sub|
#          puts "self.range = #{self.range} yielded: #{sub}"
          yield sub
        end
      end
    end
        
    def each_by_step(rng=self.range)
      initial = initial_proc.call(rng.begin)
      by_step(self.step).each do |n|
        next_begin = step_proc.call(initial,n)
        next_end   = step_proc.call(next_begin,self.length)
        if rng.respond_to?(:end) && 
           rng.end && 
           next_end > rng.end
          raise StopIteration
        end
        yield((next_begin...next_end).extend(Subrange))
      end
    end
    
    # step enumerator
    def by_step(n)
      @step_enumerator ||= Enumerator.new do |y|
        i=0
        loop do
          y << i; i+=n
        end
      end
    end
    
  end
  
end

if $0 == __FILE__

  require 'pp'
  
  x = (0..100).extend(Subrange)
  y = (2..3).extend(Subrange)
  
  pp  x.each_nth(14,1,2).to_a

  pp  y.each_nth(0.4,0.2).to_a
  
  pp  x.each_nth(14,1,2).each_nth(0.4,0.2).to_a
  
  pp  x.each_nth(20,20).each_nth(4,4).each_nth(3,1).to_a
  
end



__END__

# "every other month on third thursday at 2pm EST"
# range.each_month(2).on_thursdays(2).at_time('2:00pm EST')


def each_month(n,offset=0)
  TimeRangeEnumerator.new( self, n,
               Proc.new {|dt| dt.to_date >> offset},
               Proc.new {|dt, i| dt.to_date >> i}
             )
end

def on_wday(wd,n)
  TimeRangeEnumerator.new( self, n,
               Proc.new {|dt| (wd - dt.wday)%7 },
               Proc.new {|dt, i| dt.to_date + i*7}
             )
end

def at_time(t)
end



def each
  each_subrange do |subrange|
    yield subrange
  end
end
    
def each_subrange(&b)
  if range.respond_to?(:each_subrange)
    range.each_subrange do |subrange|
      
    end 
  else
    # get range
    
  end
end




  
  # generic time/date iteration using step enumerator below
  def each_by_step(n,step_proc,initial_proc=nil)
    initial_proc ||= lambda {|dt| dt}
    initial = initial_proc.call(self.begin)
    by_step(n).each do |step|
      next_val = step_proc.call(offset_init,step)
      if self.respond_to?(:end) && self.end && next_val > self.end
        raise StopIteration
      end
      yield(next_val)
    end
  end
  
  # step enumerator
  def by_step(n)
    Enumerator.new do |y|
      i=0
      loop do
        y << i; i+=n
      end
    end
  end


scratch/chainable_dates.rb
require 'date'
require 'time'

module DateTimeSubrange
      
  Sunday    = Date::DAYNAMES.index("Sunday")
  Monday    = Date::DAYNAMES.index("Monday")
  Tuesday   = Date::DAYNAMES.index("Tuesday")
  Wednesday = Date::DAYNAMES.index("Wednesday")
  Thursday  = Date::DAYNAMES.index("Thursday")
  Friday    = Date::DAYNAMES.index("Friday")
  Saturday  = Date::DAYNAMES.index("Saturday")
  Sun = Sunday
  Mon = Monday
  Tue = Tuesday
  Wed = Wednesday
  Thu = Thursday
  Fri = Friday
  Sat = Saturday

  January   = Date::MONTHNAMES.index("January")
  February  = Date::MONTHNAMES.index("February")
  March     = Date::MONTHNAMES.index("March")
  April     = Date::MONTHNAMES.index("April")
  May       = Date::MONTHNAMES.index("May")
  June      = Date::MONTHNAMES.index("June")
  July      = Date::MONTHNAMES.index("July")
  August    = Date::MONTHNAMES.index("August")
  September = Date::MONTHNAMES.index("September")
  October   = Date::MONTHNAMES.index("October")
  November  = Date::MONTHNAMES.index("November")
  December  = Date::MONTHNAMES.index("December")
  Jan = January
  Feb = February
  Mar = March
  Apr = April
  Jun = June
  Jul = July
  Aug = August
  Sep = September
  Oct = October
  Nov = November
  Dec = December
  
  def each_seconds(n=1, offset=0, dur=1)
    build_subrange do |s|
      s.step    = n
      s.adjust_range { |r| time_range(r) }
      s.offset       { |tm|   tm.to_time + offset }
      s.increment    { |tm,i| tm.to_time + i }
      s.span         { |tm| tm.to_time + dur }
    end
  end
  alias each_second each_seconds
  
  def each_minutes(n=1, offset=0, dur=1)
    each_seconds(n*60, offset*60, dur*60)
  end
  alias each_minute each_minutes
  
  def each_hours(n=1, offset=0, dur=1)
    each_seconds(n*60*60, offset*60*60, dur*60*60)
  end
  alias each_hour each_hours
  
  def each_days(n=1,offset=0,dur=1)
    build_subrange do |s|
      s.step    = n
      s.adjust_range  { |r| day_range(r) }
      s.offset        { |dt|   dt.to_date + offset }
      s.increment     { |dt,i| dt.to_date + i }
      s.span          { |dt| dt.to_date + dur }
    end
  end
  alias each_day each_days
  
  def each_weeks(n=1, offset=0)
    each_days(n*7, offset*7)
  end
  alias each_week each_weeks
  
  def each_wdays(wd,n=1,offset=0)
    build_subrange do |s|
      s.step    = n
      s.adjust_range  { |r| day_range(r) }
      s.offset        { |dt|   dt.to_date + (wd - dt.to_date.wday)%7 + offset*7 }
      s.increment     { |dt,i| dt.to_date + i*7 }
      s.span          { |dt| dt.to_date + 0 }
    end
  end
  alias each_wday each_wdays
    
  def each_sunday(   n=1, offset=0); each_wdays(Sun,n,offset); end
  def each_monday(   n=1, offset=0); each_wdays(Mon,n,offset); end
  def each_tuesday(  n=1, offset=0); each_wdays(Tue,n,offset); end
  def each_wednesday(n=1, offset=0); each_wdays(Wed,n,offset); end
  def each_thursday( n=1, offset=0); each_wdays(Thu,n,offset); end
  def each_friday(   n=1, offset=0); each_wdays(Fri,n,offset); end
  def each_saturday( n=1, offset=0); each_wdays(Sat,n,offset); end
  
  def each_months(n=1,offset=0,dur=1)
    build_subrange do |s|
      s.step    = n
      s.adjust_range { |r| day_range(r) }
      s.offset       { |dt|   dt.to_date >> offset }
      s.increment    { |dt,i| dt.to_date >> i }
      s.span         { |dt| dt.to_date >> dur }
    end
  end
  alias each_month each_months
  
  def each_monthnum(nmonth,n=1)
    build_subrange do |s|
      s.step   = n
      s.adjust_range { |r| day_range(r) }
      s.offset       { |dt| dt >> (nmonth - dt.month)%12 }
      s.increment    { |dt,i| dt.to_date >> i*12 }      
      s.span         { |dt| dt.to_date >> 1 }
    end
  end
  
  def each_years(n=1,offset=0,dur=1)
    build_subrange do |s|
      s.step    = n
      s.adjust_range { |r| day_range(r) }
      s.offset       { |dt|   Date.civil(dt.year + offset, dt.month, dt.day) }
      s.increment    { |dt,i| Date.civil(dt.year + i,      dt.month, dt.day) }
      s.span         { |dt|   Date.civil(dt.year + dur, dt.month, dt.day) }
    end 
  end
  alias each_year each_years
  
  
  # day of each month
  def on_day(nday,dur=0)
    build_subrange do |s|
      s.step   = 1
      s.adjust_range   { |r| day_range(r) }
      s.offset         do |dt| 
                            totdays = ((Date.civil(dt.year,dt.month,1) >> 1)-1).day
                            dt.to_date + (nday - dt.day)%totdays
                       end
      s.increment      { |dt,i| dt.to_date >> i }
      s.span           { |dt| dt.to_date + dur  }
    end
  end
  
  # time of each day
  def at_time(tm,dur=0)
    tm_p = Time.parse(tm)
    build_subrange do |s|
      s.step    = 60*60*24
      s.adjust_range    { |r| time_range(r) }        
      s.offset          do |tm| 
                             Time.new( 
                               tm.year,   tm.month, tm.day,
                               tm_p.hour, tm_p.min, tm_p.sec, tm_p.utc_offset
                             )
                        end
      s.increment       { |tm,i| tm.to_time + i }
      s.span            { |tm| tm.to_time + dur }
    end
  end
    
  # convert to date range
  # and make exclusive if not already
  def day_range(rng=self)
    if rng.respond_to?(:exclude_end?) && rng.exclude_end?
      Range.new(rng.begin.to_date, rng.end.to_date, true)
    else
      Range.new(rng.begin.to_date, rng.end.to_date + 1, true)
    end
  end
  
  # unless already a time range,
  # convert to times and make exclusive
  def time_range(rng=self)
    if rng.begin.respond_to?(:sec) && rng.end.respond_to?(:sec)
      rng.dup
    else
      Range.new(rng.begin.to_time, (rng.end.to_date + 1).to_time, true)
    end
  end
  
  def build_subrange(&builder)
    SubRangeIterator.new(self, &builder)
  end
  
  class SubRangeIterator
    include Enumerable
    include DateTimeSubrange
    
    attr_accessor :range, :step, :length
    def step;   @step   ||= 1;         end
    def length; @length ||= self.step; end
    
    def range_extensions
      @range_extensions ||= 
        class << self.range
          self.included_modules - [Kernel]
        end
    end
    
    def initialize(range)
      self.range = range
      yield self if block_given?
    end
  
    def adjust_range(&p)
      self.range_proc = p
    end
    
    def offset(&p)
      self.offset_proc = p
    end
    
    def increment(&p)
      self.step_proc = p
    end
    
    def span(&p)
      self.span_proc = p
    end
    
    def each(&b)
      if self.range.respond_to?(:each_by_step)
        self.range.each do |sub|
          each_by_step(sub, &b)
        end
      else
        each_by_step do |sub|
#          puts "self.range = #{self.range} yielded: #{sub}"
          yield sub
        end
      end
    end
        
    def each_by_step(rng=self.range)
      rng = range_proc.call(rng)
#      puts "each_by_step range: #{rng}"
      initial = offset_proc.call(rng.begin)
      by_step(self.step).each do |n|
        next_begin = step_proc.call(initial,n)
        next_end   = span_proc.call(next_begin)
        if rng.respond_to?(:cover?) && !rng.cover?(next_begin) 
          raise StopIteration
        end
        yield((next_begin...next_end).extend(*range_extensions))
      end
    end
    
    # step enumerator
    def by_step(n)
      @step_enumerator ||= Enumerator.new do |y|
        i=0
        loop do
          y << i; i+=n
        end
      end
    end
  
    private
    
    attr_accessor :offset_proc, :step_proc, :span_proc, :range_proc
    def range_proc
      @range_proc ||= lambda {|r| r}
    end
  end
  
end


if $0 == __FILE__

  require 'pp'
  
  x = (Date.civil(2012,1,1)..Date.civil(2012,3,31)).extend(DateTimeSubrange)

  
  puts "Each Sunday at 11am UTC"
  pp x.each_month.each_wday(0).at_time('11:00 UTC').to_a
  puts
  
  puts "Every second Monday in every other month"
  pp x.each_month(2).each_wday(1,2,1).to_a
  puts
  
  puts "Each 15th of the month"
  pp x.each_month.on_day(15).to_a
  puts

  puts "Each February"
  pp x.each_monthnum(2).to_a
  
end


scratch/temporal_enumerable.rb
require 'time'
require 'date'

module TemporalEnumerable

  Sunday    = Date::DAYNAMES.index("Sunday")
  Monday    = Date::DAYNAMES.index("Monday")
  Tuesday   = Date::DAYNAMES.index("Tuesday")
  Wednesday = Date::DAYNAMES.index("Wednesday")
  Thursday  = Date::DAYNAMES.index("Thursday")
  Friday    = Date::DAYNAMES.index("Friday")
  Saturday  = Date::DAYNAMES.index("Saturday")
  Sun = Sunday
  Mon = Monday
  Tue = Tuesday
  Wed = Wednesday
  Thu = Thursday
  Fri = Friday
  Sat = Saturday
  
  #---------------- step by seconds
  def each_seconds(n=1, off=0, &b)
    t0 = self.begin.to_time
    each_by_step(n, off, t0, lambda{|t,i| t+i}, &b)
  end
  alias each_second each_seconds
  
  def second(n)
    time = nil
    each_second(1,n) {|t| time = t; break}
    time
  end
  
  def each_minutes(n=1, off=0, &b)
    each_seconds(n*60, off*60, &b)
  end
  alias each_minute each_minutes
  
  def minute(n)
    time =nil
    each_minute(1,n) {|t| time = t; break}
    time
  end
  
  def each_hours(n=1, off=0, &b)
    each_seconds(n*60*60, off*60*60, &b)
  end
  alias each_hour each_hours
  
  def hour(n)
    time = nil
    each_hour(1,n) {|t| time = t; break}
    time
  end
  
  # -----
  
  # step by days
  def each_days(n=1, off=0, &b)
    d0 = self.begin.to_date
    each_by_step(n, off, d0, lambda{|t,i| t+i}, &b)
  end
  alias each_day each_days
  
  def day(n)
    dt = nil
    each_day(1,n) {|t| dt = t; break}
    dt
  end
  
  def each_weeks(n=1, off=0, &b)
    each_days(n*7, off*7, &b)
  end
  alias each_week each_weeks
  
  def week(n)
    dt = nil
    each_week(1,n) {|t| dt = t; break}
    dt
  end
  
  # -----
  
  # step by weeks on a specific weekday
  def each_wdays(wd, n=1, off=0, &b)
    d0 = self.begin.to_date
    d0+=((wd-d0.wday)%7)
    each_by_step(n*7, off*7, d0, lambda {|d,i| d + i}, &b)
  end
  alias each_wday each_wdays
  
  def wday(wd, n)
    dt = nil
    each_wday(wd,1,n) {|t| dt = t; break}
    dt
  end
  
  def each_sunday(   n=1, off=0, &b); each_wdays(Sun,n,off,&b); end
  def each_monday(   n=1, off=0, &b); each_wdays(Mon,n,off,&b); end
  def each_tuesday(  n=1, off=0, &b); each_wdays(Tue,n,off,&b); end
  def each_wednesday(n=1, off=0, &b); each_wdays(Wed,n,off,&b); end
  def each_thursday( n=1, off=0, &b); each_wdays(Thu,n,off,&b); end
  def each_friday(   n=1, off=0, &b); each_wdays(Fri,n,off,&b); end
  def each_saturday( n=1, off=0, &b); each_wdays(Sat,n,off,&b); end
  
  def sunday(   n); wday(Sun,n); end
  def monday(   n); wday(Mon,n); end
  def tuesday(  n); wday(Tue,n); end
  def wednesday(n); wday(Wed,n); end
  def thursday( n); wday(Thu,n); end
  def friday(   n); wday(Fri,n); end
  def saturday( n); wday(Sat,n); end

  # -----
  
  # step by months
  def each_months(n=1, off=0, &b)
    d0 = self.begin.to_date
    each_by_step(n, off, d0, lambda {|d,i| d >> i}, &b)
  end
  alias each_month each_months
  
  def month(n)
    dt = nil
    each_month(1,n) {|t| dt = t; break}
    dt
  end
  
  # -----
  
  # step by years
  def each_years(n=1, off=0, &b)
    d0 = self.begin.to_date
    each_by_step( n, off, d0, 
                  lambda {|d,i| Date.civil(d.year + i,d.month,d.day)},
                  &b
                )  
  end
  alias each_year each_years
  
  def year(n)
    dt = nil
    each_year(1,n) {|t| dt = t; break}
    dt
  end
  
  # -----
  
  # date/time range intersection
  
  def intersection(other, precision=0)
    if ( self.begin.respond_to?(:seconds) &&  self.end.respond_to?(:seconds)) ||
       (other.begin.respond_to?(:seconds) && other.end.respond_to?(:seconds))
      time_intersection(other,precision)

    elsif ( self.begin.respond_to?(:days) &&  self.end.respond_to?(:days)) ||
          (other.begin.respond_to?(:days) && other.end.respond_to?(:days))
      date_intersection(other,precision)
    
    else
      raise TypeError
    end
  end
  
  def & (other)
    intersection(other)
  end
  
  def time_intersection(other, precision=0)
    trange1 = fuzz_time_range(self, precision)
    trange2 = fuzz_time_range(other,precision)
    b = [trange1.begin,trange2.begin].max
    e = [trange1.end  ,trange2.end  ].min
    if e < b
      nil..nil
    else
      Range.new(b, e, self.exclude_end? && precision==0)
    end
  end
  
  def date_intersection(other, precision=0)
    drange1 = fuzz_date_range(self, precision)
    drange2 = fuzz_date_range(other,precision)
    b = [drange1.begin,drange2.begin].max
    e = [drange1.end  ,drange2.end  ].min
    if e < b
      nil..nil    # not sure this is the best way
    else
      Range.new(b, e, self.exclude_end? && precision==0).extend(self.class)
    end 
  end
  
  # probably should be private
  def fuzz_time_range(range,precision)
    if range.respond_to?(:begin) && range.respond_to?(:end)
      Range.new(range.begin.to_time - precision,
                range.end.to_time   + precision,
                range.exclude_end?
               )
    else
      Range.new(range.to_time - precision,
                range.to_time + precision,
                false
               )
    end    
  end

  # probably should be private
  def fuzz_date_range(range,precision)
    if range.respond_to?(:begin) && range.respond_to?(:end)
      Range.new(range.begin.to_date - precision,
                range.end.to_date   + precision,
                range.exclude_end?
               )
    else
      Range.new(range.to_date - precision,
                range.to_date + precision,
                false
               )
    end    
  end
  
  private
  
  # generic time/date iteration using step enumerator below
  def each_by_step(n,offset,initial,step_proc)
    accum = []
    offset_init = step_proc.call(initial,offset)
    by_step(n).each do |step|
      next_val = step_proc.call(offset_init,step)
      if self.respond_to?(:end) && self.end && next_val > self.end
        raise StopIteration
      end
      block_given? ? yield(next_val) : accum << next_val
    end
    block_given? ? nil : accum
  end
  
  # step enumerator
  def by_step(n)
    Enumerator.new do |y|
      i=0
      loop do
        y << i; i+=n
      end
    end
  end
  
end
 
  
class Schedule
  include Enumerable
  
  attr_reader :events
  def events; @events ||= []; end
  
  def add_event(event, time=nil, duration={}, &expr)
    e = Event.new(event,time,duration,&expr)
    events << e
    e
  end

  # Probably, we should set range as a state of the schedule
  # rather than pass it in to each of these methods
  
  def each(range,&b)
    events.each(range,&b)
  end
  
  # each event with at least one date in the passed range
  # normally you would not use this unless you just need a list of event names
  # instead use #in for more flexibility
  def events_in(range)
    events.reject {|event| 
      event.each(range).count == 0
    }
  end
  
  # schedule records for the passed range
  def in(range)
    events.inject([]) {|memo,event|
      dts = event.each(range).to_a
      unless dts.empty?
        rec = schedule_record.new
        rec.event = event
        rec.dates = dts
        memo << rec
      end
      memo
    }
  end
  
  # TODO
  def conflicts_in(range)
  end
  
  def schedule_record
    @schedule_record ||= Struct.new(:event, :dates)
  end
  
  # TODO: events need duration or internal time range
  class Event
    include Enumerable
    
    attr_accessor :name
    
    def initialize(name, time=nil, mods={}, &cond)
      self.name      = name
      self.precision = mods.delete(:precision) || 0
      schedule(&cond)
    end
    
    def or(time=nil, mods={}, &cond)
      cond ||= lambda {|range| 
                  subrange = event_range(time,mods) 
                  range.intersection(subrange, self.precision)
               }
      stack << cond
      self
    end
        
    def schedule(time=nil, mods={}, &cond)
      init_stack; self.or(time, mods, &cond)
    end
    
    def each(range,&b)
      stack.inject([]) do |results,cond|
        results | cond.call(range)
      end.each(&b)
    end
    alias each_date each
        
    def include_date?(dt, range)
      !!stack.find {|cond| cond.call(range).include?(dt)}
    end
    alias include_time? include_date?
    
    def to_s
      "#{name} (#{stack.size} conditions)"
    end
    
    def event_range(time,mods)
      dur = 
    end
    
    
    private
    
    attr_reader :stack
    def stack; @stack ||= []; end
    def init_stack; @stack = nil; stack; end
   
    
  end
  
end

if $0 == __FILE__

  require 'pp'
  
  container = Struct.new(:begin, :end).send(:include, TemporalEnumerable)
  
  dt = container.new(Time.now)
  
  puts 'every hour until the next midnight'
  dt.each_hour do |t| break if t.day != dt.begin.day && t.hour == 0; puts t end
  puts
  
  puts 'every hour starting 2 hours from now, until the next midnight'
  dt.each_hours(1,2) do |t| break if t.day != dt.begin.day && t.hour ==0; puts t end
  puts
  
  puts 'the fourth hour from now'
  puts dt.hour(4)
  puts
  
  puts 'every 2 weeks from now until March'
  dt.each_weeks(2) do |t| break if t.month == 3; puts t end
  puts 
  
  puts 'every week starting 2 weeks from now until May'
  dt.each_weeks(2,2) do |t| break if t.month == 5; puts t end
  puts
  
  dt2 = container.new(Date.today, Date.civil(Date.today.year + 10, Date.today.month, Date.today.day))
  
  puts 'every 3 months until 10 years from today'
  dt2.each_months(3) do |d| puts d end
  puts
  
  puts 'every year until 10 years from today, starting next year'
  dt2.each_year(1,1) do |d| puts d.year end
  puts
  
  puts 'first Wed in March 2012'
  # obviously you'd want a more useful container than this that just takes a year and month
  dt3 = container.new(Date.civil(2012,3,1), Date.civil(2012,3,31))
  puts dt3.each_wednesday.first
  puts
  
  puts 'second and fourth Friday in March-April 2012'
  dt4 = container.new(Date.civil(2012,4,1), Date.civil(2012,4,30))
  puts dt3.each_friday(2,1) | dt4.each_friday(2,1)
  puts
  
  # using a range
  puts 'first Wed in April 2012, using a range'
  dt5 = (Date.civil(2012,4,1)..Date.civil(2012,4,30)).extend(TemporalEnumerable)
  puts dt5.each_wednesday.first
  puts
  
  puts 'scheduling a meeting every other wednesday and the 3rd thursday and every thursday following'
  puts 'and two appointments, one of which is in the next month'
  puts 'here is the schedule of events for April 2012'
  sched = Schedule.new
  evt = Schedule::Event.new("meeting") do |range| 
           range.each_wednesday(2) 
         end.or do |range| 
           range.each_thursday(1,2)
         end
  sched.events << evt
  sched.add_event("appointment", Time.utc(2012,4,15,14,30))
  sched.add_event("next month", Time.utc(2012,5,15,14,30))
 
  pp sched.in(dt5)
  puts
  
  puts 'in April, meeting events'
  pp sched.events_in(dt5).select {|evt| evt.name == 'meeting'} 
  puts
  
  puts 'in April, is there an event Wed April 4, 2012?'
  puts sched.events.first.include_date?(Date.civil(2012,4,4), dt5)
  puts 

  dt6 = (Date.civil(2012,5,1)..Date.civil(2012,5,31)).extend(TemporalEnumerable)
  
  puts 'in May, is there an event Wed May 2, 2012?'  
  puts sched.events.first.include_date?(Date.civil(2012,5,2), dt6)
  puts 
  
  puts 'in May, is there an event May 15, 2012 at 14:30?'  
  puts sched.events.first.include_time?(Time.utc(2012,5,15,14,30), dt6)
  puts 
  
end
  
